[DCP][state_dict] Implement pin_memory and shared_memory copy for _offload_state_dict_to_cpu

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120378

**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory or share_memory version of `cpu_offload_state_dict`.

This PR also adds `_create_cpu_state_dict` to allow users to easily create a pin_memory or share_memory cpu state_dict.

**Performance improvement**
```
# The micro-benchmark has a source state_dict with 150 tensors, and each tensor is 50MB. 
# The micro-benchmark is run on a H100 machine with PCIe 5

cpu_state_dict_2 = _create_cpu_state_dict(state_dict, pin_memory=True)
cpu_state_dict_3 = _create_cpu_state_dict(state_dict, share_memory=True)

# GPU->CPU memory: 4.6556 seconds
cpu_state_dict = _offload_state_dict_to_cpu(state_dict)

# GPU->pin memory: 0.1566 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)

# GPU->shared memory: 0.5509 seconds (variation is quite large)
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_3)

# GPU->pin memory->shared memory: 0.2550 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)
_offload_state_dict_to_cpu(cpu_state_dict_2, cpu_offload_state_dict=cpu_state_dict_3)
```


Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @LucasLLC

fegin committed:
[DCP][state_dict] Implement pin_memory copy for _offload_state_dict_to_cpu

This may speed up the cpu offloading.

Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

[ghstack-poisoned]

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory `cpu_offload_state_dict`, which a micro-benchmark show the performance improvement. 

This PR also adds `_create_pin_state_dict` to allow users to easily create a pin_memory cpu state_dict.

**Performance improvement**
A micro-benchmark with 150 tensors state_dict, each tensor is 15MB, shows ~ 8X performance improvement, from 1.6 seconds to 0.21 seconds. The micro-benchmark was run on a A100 GPU.

Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory `cpu_offload_state_dict`, which a micro-benchmark show the performance improvement. 

This PR also adds `_create_pin_state_dict` to allow users to easily create a pin_memory cpu state_dict.

**Performance improvement**
A micro-benchmark with 150 tensors state_dict, each tensor is 15MB, shows ~ 8X performance improvement, from 1.6 seconds to 0.21 seconds. The micro-benchmark was run on a A100 GPU.

Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory `cpu_offload_state_dict`, which a micro-benchmark show the performance improvement. 

This PR also adds `_create_pin_state_dict` to allow users to easily create a pin_memory cpu state_dict.

**Performance improvement**
A micro-benchmark with 150 tensors state_dict, each tensor is 15MB, shows ~ 8X performance improvement, from 1.6 seconds to 0.21 seconds. The micro-benchmark was run on a A100 GPU.

Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -54,56 +64,80 @@ def _iterate_state_dict(
         ret = iter_object
     elif isinstance(iter_object, dict):
+        if companion_obj is None:
+            companion_obj = {k: None for k in iter_object}
```
```suggestion
            companion_obj = dict.fromkeys(iter_object}
```

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory `cpu_offload_state_dict`, which a micro-benchmark show the performance improvement. 

This PR also adds `_create_pin_state_dict` to allow users to easily create a pin_memory cpu state_dict.

**Performance improvement**
A micro-benchmark with 150 tensors state_dict, each tensor is 15MB, shows ~ 8X performance improvement, from 1.6 seconds to 0.21 seconds. The micro-benchmark was run on a A100 GPU.

Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

yifuwang commented:
According to the documentation of [cudaHostAlloc](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gb65da58f444e7230d3322b6126bb4902):
> Allocating excessive amounts of pinned memory may degrade system performance, since it reduces the amount of memory available to the system for paging. As a result, this function is best used sparingly to allocate staging areas for data exchange between host and device.

Allocating too much pinned memory can lead to swapping or even OOM'ing. The threshold probably depend on the host's free and rss memory. We need to be a bit careful since state dict can be quite large (also x8 from the host's perspective).

Maybe consider measuring the system's free/rss memory and fallback to unpinned memory if the usage could cause problems.

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory `cpu_offload_state_dict`, which a micro-benchmark show the performance improvement. 

This PR also adds `_create_pin_state_dict` to allow users to easily create a pin_memory cpu state_dict.

**Performance improvement**
A micro-benchmark with 150 tensors state_dict, each tensor is 15MB, shows ~ 8X performance improvement, from 1.6 seconds to 0.21 seconds. The micro-benchmark was run on a A100 GPU.

Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin commented:
@yifuwang Thanks for the comment. Agree this may not always be the best solution, and can be slower sometimes. This is an opt-in feature and requires the user to specifically allocates the pin memory beforehand and maintains the pin memory. So it is not suitable for the API to do such a fallback. The user has to do the work and decide whether there are enough memory.

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory `cpu_offload_state_dict`, which a micro-benchmark show the performance improvement. 

This PR also adds `_create_pin_state_dict` to allow users to easily create a pin_memory cpu state_dict.

**Performance improvement**
A micro-benchmark with 150 tensors state_dict, each tensor is 15MB, shows ~ 8X performance improvement, from 1.6 seconds to 0.21 seconds. The micro-benchmark was run on a A100 GPU.

Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

jspark1105 commented:
Nice! In the example in the summary
```
# GPU->shared memory->pin memory: 0.2550 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)
_offload_state_dict_to_cpu(cpu_state_dict_2, cpu_offload_state_dict=cpu_state_dict_3)
```
"GPU->shared memory->pin memory"
supposed to be "GPU->pin memory->shared memory"?

fegin commented:
@jspark1105 you are right. Updated the summary.

LucasLLC commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
     return tensor
 
 
+class CompanionMismatch(Exception):
```
Any reason we can't use a value error here?

fegin replied:
We need a customized exception to distinguish mismatched error from the actual error. Since the mismatched error is not an actual error, we will silently ignore the error. So if we use ValueError, we may accidentally ignore a real error.

LucasLLC commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
     pg: Optional[dist.ProcessGroup] = None,
     device: Optional[torch.device] = None,
     cpu_offload: bool = False,
+    companion_obj: Any = None,
```
We can take care of this in a different PR (actually I'm happy to take it up), but a docstring for this function would be really great. Mostly because `campanion_obj` is a little difficult to interpret.

fegin replied:
:/ yup, the document needs some improvement.

LucasLLC commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
+        elif not isinstance(companion_obj, dict) or set(companion_obj.keys()) != set(
+            iter_object.keys()
+        ):
+            raise CompanionMismatch()
```
nit: we could consider adding something like:

 `f"{type(companion_obj)} {set(companion_obj.keys()) - set(iter_object.keys()) }"` to the exception to provide some insight.

fegin replied:
hmm, I'm not sure if the error message is going to help since we just silently ignore the CompanionMismatch error. But it doesn't hurt to print out the keys later I think.

LucasLLC commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
+        ):
+            raise CompanionMismatch()
+
+        assert isinstance(companion_obj, dict)
```
nit: assertion is unreachable

fegin replied:
Thanks for catching this

LucasLLC commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -50,60 +60,97 @@ def _all_gather_sharded_tensor(
     ):
         ret = iter_object
     elif isinstance(iter_object, dict):
+        if companion_obj is None:
```
I think I may be missing something but why create a campanion obj if it is None? Can't we just skip the comparison in that case?

fegin replied:
This is just a trick to make the comparison easily. Since we will unroll the companion object as well. So instead of check whether it is None and pass None else passing companion[key], I create a dictionary full of None so that I can also write `companion[key]`. But this may be too implicit. I can change this.

LucasLLC commented on torch/distributed/_state_dict_utils.py:
```diff
@@ -185,26 +236,152 @@ def dtensor_func(value, pg, device):
+    return ret
+
+
+def _check_state_dict_similarity(
```
Would it be difficult to always do this when we have a companion obj in `_iterate` ? That way users wouldn't need to call this method before using pinned memory, etc.

fegin replied:
It's not difficult but I'm not sure if it is better. If we check in the offload_cpu call, users have to do EAFP style, using a try-catch to see if there are errors. However, in this case, if the exception happens, users will need to do the mitigation and retry offload_cpu call. The error is not an actual error as it is possible that state_dict changes. If users just use LBYL style, using a if statement to check, it is simpler IMHO. No retry just a check anyway.

LucasLLC reviewed (commented):
Great work! This is an awesome performance improvement and it came out super quick!

LucasLLC reviewed (approved):
LGTM!

One small nit: I do think your suggestion around making `companion_obj` none would read better :)

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory and shared_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory or share_memory version of `cpu_offload_state_dict`.

This PR also adds `_create_cpu_state_dict` to allow users to easily create a pin_memory or share_memory cpu state_dict.

**Performance improvement**
```
# The micro-benchmark has a source state_dict with 150 tensors, and each tensor is 50MB. 
# The micro-benchmark is run on a H100 machine with PCIe 5

cpu_state_dict_2 = _create_cpu_state_dict(state_dict, pin_memory=True)
cpu_state_dict_3 = _create_cpu_state_dict(state_dict, share_memory=True)

# GPU->CPU memory: 4.6556 seconds
cpu_state_dict = _offload_state_dict_to_cpu(state_dict)

# GPU->pin memory: 0.1566 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)

# GPU->shared memory: 0.5509 seconds (variation is quite large)
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_3)

# GPU->pin memory->shared memory: 0.2550 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)
_offload_state_dict_to_cpu(cpu_state_dict_2, cpu_offload_state_dict=cpu_state_dict_3)
```


Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory and shared_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory or share_memory version of `cpu_offload_state_dict`.

This PR also adds `_create_cpu_state_dict` to allow users to easily create a pin_memory or share_memory cpu state_dict.

**Performance improvement**
```
# The micro-benchmark has a source state_dict with 150 tensors, and each tensor is 50MB. 
# The micro-benchmark is run on a H100 machine with PCIe 5

cpu_state_dict_2 = _create_cpu_state_dict(state_dict, pin_memory=True)
cpu_state_dict_3 = _create_cpu_state_dict(state_dict, share_memory=True)

# GPU->CPU memory: 4.6556 seconds
cpu_state_dict = _offload_state_dict_to_cpu(state_dict)

# GPU->pin memory: 0.1566 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)

# GPU->shared memory: 0.5509 seconds (variation is quite large)
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_3)

# GPU->pin memory->shared memory: 0.2550 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)
_offload_state_dict_to_cpu(cpu_state_dict_2, cpu_offload_state_dict=cpu_state_dict_3)
```


Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]

fegin committed:
Update on "[DCP][state_dict] Implement pin_memory and shared_memory copy for _offload_state_dict_to_cpu"


**Summary**
This PR extend `_offload_state_dict_to_cpu` to accept a `cpu_offload_state_dict` argument. If `cpu_offload_state_dict` is not None, `_offload_state_dict_to_cpu` will use `copy_` to copy the GPU data to the CPU tensors. This allows users to pass a pin_memory or share_memory version of `cpu_offload_state_dict`.

This PR also adds `_create_cpu_state_dict` to allow users to easily create a pin_memory or share_memory cpu state_dict.

**Performance improvement**
```
# The micro-benchmark has a source state_dict with 150 tensors, and each tensor is 50MB. 
# The micro-benchmark is run on a H100 machine with PCIe 5

cpu_state_dict_2 = _create_cpu_state_dict(state_dict, pin_memory=True)
cpu_state_dict_3 = _create_cpu_state_dict(state_dict, share_memory=True)

# GPU->CPU memory: 4.6556 seconds
cpu_state_dict = _offload_state_dict_to_cpu(state_dict)

# GPU->pin memory: 0.1566 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)

# GPU->shared memory: 0.5509 seconds (variation is quite large)
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_3)

# GPU->pin memory->shared memory: 0.2550 seconds
_offload_state_dict_to_cpu(state_dict, cpu_offload_state_dict=cpu_state_dict_2)
_offload_state_dict_to_cpu(cpu_state_dict_2, cpu_offload_state_dict=cpu_state_dict_3)
```


Differential Revision: [D54045845](https://our.internmc.facebook.com/intern/diff/D54045845/)

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 LucasLLC

[ghstack-poisoned]
